### PR TITLE
Add unit test for FontCascade::characterRangeCodePath non surrogated ranges

### PR DESCRIPTION
**Pull request by vitorroriz**
#### 64ce55deb8240e7b2078379961046fd26e490754
<pre>
Add unit test for FontCascade::characterRangeCodePath non surrogated ranges
<a href="https://bugs.webkit.org/show_bug.cgi?id=276494">https://bugs.webkit.org/show_bug.cgi?id=276494</a>
<a href="https://rdar.apple.com/131478343">rdar://131478343</a>

Reviewed by Ryan Reno.

Adding unit tests for the non-surrogated ranges of FontCascade::characterRangeCodePath

This was inspired by the following Blink patch:
<a href="https://codereview.chromium.org/18949006/">https://codereview.chromium.org/18949006/</a>

* Source/WebCore/platform/graphics/FontCascade.h:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebCore/FontCascade.cpp: Added.
(TestWebKitAPI::testCodePath):
(TestWebKitAPI::testCodePathRange):
(TestWebKitAPI::TEST(FontCascadeTest, characterRangeCodePath_NonSurrogates)):

Canonical link: <a href="https://commits.webkit.org/280888@main">https://commits.webkit.org/280888@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/a6d6e03a2f328b60a8f5a7d5c4d7c4d963e66bc6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/57931 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/37259 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/10407 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/61553 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/8376 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/60059 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/44895 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/8564 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46952 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5970 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59961 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34942 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/50071 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27779 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/31708 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/7365 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/7380 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/53655 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/7634 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/63240 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1845 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/7704 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/54173 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1852 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/50082 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/54308 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12816 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/1576 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/33088 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/34174 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/35258 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33919 "Built successfully") | | |

### CODE REVIEW

File: Tools/TestWebKitAPI/Tests/WebCore/FontCascade.cpp [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+#include <WebCore/FontCascade.h>
+
+namespace TestWebKitAPI {
+
+using CodePath = WebCore::FontCascade::CodePath;
+using FontCascade = WebCore::FontCascade;
+
+static void testCodePath(UChar codePoint, CodePath codePath)
+{
+    UChar target[1] = { codePoint };
```
Comment (mdubet) [inline Tools/TestWebKitAPI/Tests/WebCore/FontCascade.cpp]: `std::array` ?

File: Tools/TestWebKitAPI/Tests/WebCore/FontCascade.cpp [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include <WebCore/FontCascade.h>
+
+namespace TestWebKitAPI {
+
+using CodePath = WebCore::FontCascade::CodePath;
```
Comment (rreno) [inline Tools/TestWebKitAPI/Tests/WebCore/FontCascade.cpp]: nit: you should be able to just do `using WebCore::FontCascade::CodePath;` (and likewise for the next one).

File: Tools/TestWebKitAPI/Tests/WebCore/FontCascade.cpp [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+static void testCodePathRange(CodePathRange range)
+{
+    ASSERT(range.start > 0);
+    UChar below[1] = { static_cast<UChar>(range.start - 1) };
+    UChar start[1] = { range.start };
+    UChar middle[1] = { static_cast<UChar>((range.start + range.end) / 2) };
+    UChar end[1] = { range.end };
+    UChar above[1] = { static_cast<UChar>(range.end + 1) };
+
+    EXPECT_EQ(range.belowPath, WebCore::FontCascade::characterRangeCodePath(std::span<UChar>(below))) << "below: " << std::hex << static_cast<int>(below[0]);
```
Comment (rreno) [inline Tools/TestWebKitAPI/Tests/WebCore/FontCascade.cpp]: nit: you probably don't need the `WebCore::` prefix anymore with the `using` statements above.

File: Tools/TestWebKitAPI/Tests/WebCore/FontCascade.cpp [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    // Devanagari, Bengali, Gurmukhi, Gujarati, Oriya, Tamil, Telugu, Kannada,
+    // Malayalam, Sinhala, Thai, Lao, Tibetan, Myanmar
+    testCodePathRange({ 0x0600, 0x109F, CodePath::Complex });
+    // U+1100 through U+11FF Hangul Jamo (only Ancient Korean should be left here if you precompose;
+    // Modern Korean will be precomposed as a result of step A)
+    testCodePathRange({ 0x1100, 0x11FF, CodePath::Complex });
+    // U+135D through U+135F Ethiopic combining marks
+    testCodePathRange({ 0x135D, 0x135F, CodePath::Complex });
+    // U+1780 through U+18AF Tagalog, Hanunoo, Buhid, Taghanwa,Khmer, Mongolian
+    testCodePathRange({ 0x1700, 0x18AF, CodePath::Complex });
```
Comment (rreno) [inline Tools/TestWebKitAPI/Tests/WebCore/FontCascade.cpp:85]: Should this be 0x1780 or is the comment wrong?

File: Tools/TestWebKitAPI/Tests/WebCore/FontCascade.cpp (hunk 3 above)
Comment (vitorroriz) [reply Tools/TestWebKitAPI/Tests/WebCore/FontCascade.cpp:57]: That doesn't work: `using declaration cannot refer to class member`. I needed the alias.

File: Tools/TestWebKitAPI/Tests/WebCore/FontCascade.cpp (hunk 2 above)
Comment (vitorroriz) [reply Tools/TestWebKitAPI/Tests/WebCore/FontCascade.cpp]: That doesn't work: using declaration cannot refer to class member. I needed the alias.

Comment (rreno) [reply Tools/TestWebKitAPI/Tests/WebCore/FontCascade.cpp]: Ah didn't realize it was a class member. Alias is fine, I don't feel all that strongly about how `using` is used.

**Review by rreno — APPROVED** [review]

File: Source/WebCore/platform/graphics/FontCascade.cpp [hunk 5]
```diff
@@ -707,7 +707,7 @@ FontCascade::CodePath FontCascade::characterRangeCodePath(std::span<const UChar>
         if (c <= 0x135F)
             return CodePath::Complex;
 
-        if (c < 0x1700) // U+1780 through U+18AF Tagalog, Hanunoo, Buhid, Taghanwa,Khmer, Mongolian
```
Comment (stwrt) [inline Source/WebCore/platform/graphics/FontCascade.cpp]: This should be a separate patch. I am not sure this is needed for adding a unit test.

Comment (vitorroriz) [reply Source/WebCore/platform/graphics/FontCascade.cpp]: I'm just fixing the comment as I'm copying it to the unit test but I can do that on another point...

File: Source/WebCore/platform/graphics/FontCascade.cpp [hunk 6]
```diff
@@ -880,14 +836,22 @@ FontCascade::CodePath FontCascade::characterRangeCodePath(std::span<const UChar>
             continue;
         }
 
-        if (c < 0xFE00) // U+FE00 through U+FE0F Unicode variation selectors
-            continue;
-        if (c <= 0xFE0F)
-            return CodePath::Complex;
+        // Search for complex cases
+        auto* boundingCharacter = WTF::approximateBinarySearch<const UChar, const UChar>(complexCodePathRanges, complexCodePathRangesSize, c, keyExtractorUChar);
```
Comment (mdubet) [inline Source/WebCore/platform/graphics/FontCascade.cpp]: Would a map be faster here ? We have `SortedArrayMap`

Comment (vitorroriz) [reply Source/WebCore/platform/graphics/FontCascade.cpp]: Two different PRs got merged by accident.

Comment (rreno) [reply Source/WebCore/platform/graphics/FontCascade.cpp]: I was gonna say I didn't approve all this!
